storage/innobase/handler/ha_innodb.cc

```
@@ -15197,6 +15197,117 @@ ha_innobase::optimize(
+                        table->name.m_name);
+    corruption= true;
+    goto err_exit;
+  }
+
+  if (!page_is_leaf(root_block->page.frame))
+  {
+    push_warning_printf(thd, Sql_condition::WARN_LEVEL_WARN, ER_NOT_KEYFILE,
+			"InnoDB: Non leaf page exists for sequence table %s.",
+                        table->name.m_name);
+    corruption= true;
+    goto err_exit;
+  }
+
+  if (page_get_n_recs(root_block->page.frame) != 1)
+  {
+    push_warning_printf(thd, Sql_condition::WARN_LEVEL_WARN, ER_NOT_KEYFILE,
+			"InnoDB: Should have only one record in sequence "
+                        "table %s. But it has %d records.", table->name.m_name,
+                        page_get_n_recs(root_block->page.frame));
+    corruption= true;
+    goto err_exit;
+  }
+
+  rec= page_rec_get_next(page_get_infimum_rec(root_block->page.frame));
+
+  if (rec_get_deleted_flag(rec, dict_table_is_comp(table)))
+  {
+    corruption= true;
+    push_warning_printf(thd, Sql_condition::WARN_LEVEL_WARN, ER_NOT_KEYFILE,
+			"InnoDB: Encountered delete marked record in sequence "
+                        "table %s.", table->name.m_name);
+     goto err_exit;
+  }
+
+  if (trx_read_trx_id(rec + clust_index->trx_id_offset) != 0 ||
+      trx_read_roll_ptr(rec + clust_index->trx_id_offset + DATA_TRX_ID_LEN) !=
+        roll_ptr_t{1} << ROLL_PTR_INSERT_FLAG_POS)
+  {
+    corruption= true;
+    push_warning_printf(thd, Sql_condition::WARN_LEVEL_WARN, ER_NOT_KEYFILE,
+			"InnoDB: Record in sequence table %s is corrupted.",
+                        table->name.m_name);
+    goto err_exit;
+  }
+
+  if (!table->no_rollback())
+  {
+    push_warning_printf(thd, Sql_condition::WARN_LEVEL_WARN, ER_NOT_KEYFILE,
+			"InnoDB: Sequence table %s has ROLLBACK enabled.",
+                        table->name.m_name);
+    corruption= true;
+  }
+err_exit:
+  mtr.commit();
+func_exit:
+  if (corruption)
+  {
+    dict_set_corrupted(clust_index, "Table corruption");
+    return false;
```

Can you please add a case where we check the sequence operation like next_val() after marking it corrupt. We should return a right error message. This would probably need you to alter it back to sequence and use the old frm file.